### ISSUE TITLE
Misc. bug: [server] Using q8_0 for KV cache reduces performance when also using a draft model

### ISSUE THREAD

**Issue by mostlygeek** [ggerganov/llama.cpp#10552]
### Name and Version

```
$ /mnt/nvme/llama-server/llama-server-be0e35 --version
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:    no
ggml_cuda_init: GGML_CUDA_FORCE_CUBLAS: no
ggml_cuda_init: found 4 CUDA devices:
  Device 0: NVIDIA GeForce RTX 3090, compute capability 8.6, VMM: yes
  Device 1: Tesla P40, compute capability 6.1, VMM: yes
  Device 2: Tesla P40, compute capability 6.1, VMM: yes
  Device 3: Tesla P40, compute capability 6.1, VMM: yes
version: 4187 (be0e350c)
built with cc (Ubuntu 13.2.0-23ubuntu4) 13.2.0 for x86_64-linux-gnu
```

### Operating systems

Linux

### Which llama.cpp modules do you know to be affected?

llama-server

### Problem description & steps to reproduce

There is a pretty consistent 16% tokens/second performance drop when using `--cache-type-k q8_0 --cache-type-v q8_0` with a draft model. It doesn't happen if I don't use a draft model. 

| model | python | typescript | swift | 
|-------|--------|------------|-------|
| qwen-coder-32b-q4 | 79.9 | 54.48 | 46.67 |
| qwen-coder-32b-q4-kv | 66.60 (-16.6%) | 45.27 (-16%) | 39.24 (-15.9%) | 

The test I am using is to prompt for a snake game to be written in python, typescript and swift. 

```
for model in "qwen-coder-32b-q4" "qwen-coder-32b-q4-kv"; do 
    for lang in "python" "typescript" "swift"; do
        echo "Generating Snake Game in $lang using $model"
        curl -s --url http://localhost:8080/v1/chat/completions -d "{\"messages\": [{\"role\": \"system\", \"content\": \"you only write code.\"}, {\"role\": \"user\", \"content\": \"write snake game in $lang\"}], \"temperature\": 0.1, \"model\":\"$model\"}" > /dev/null
    done
done
```

Here is my [llama-swap](https://github.com/mostlygeek/llama-swap) configuration for the models above. The only difference is the addition of the cache-type flags:

```yaml
models:
  "qwen-coder-32b-q4":
    # main model on 3090, draft on P40 #1
    #
    # gist results: python: 79.97 tps, typescript: 54.48 tps, swift: 46.67 tps
    cmd: >
      /mnt/nvme/llama-server/llama-server-be0e35
      --host 127.0.0.1 --port 9503
      --flash-attn --metrics
      --slots
      --model /mnt/nvme/models/Qwen2.5-Coder-32B-Instruct-Q4_K_M.gguf
      -ngl 99
      --ctx-size 19000
      --model-draft /mnt/nvme/models/Qwen2.5-Coder-0.5B-Instruct-Q8_0.gguf
      -ngld 99
      --draft-max 16
      --draft-min 4
      --draft-p-min 0.4
      --device CUDA0
      --device-draft CUDA1
    proxy: "http://127.0.0.1:9503"

  "qwen-coder-32b-q4-kv":
    # main model on 3090, draft on P40 #1
    #
    # gist results: python: 66.60, typescript 45.27, swift 39.24
    cmd: >
      /mnt/nvme/llama-server/llama-server-be0e35
      --host 127.0.0.1 --port 9503
      --flash-attn --metrics
      --slots
      --model /mnt/nvme/models/Qwen2.5-Coder-32B-Instruct-Q4_K_M.gguf
      -ngl 99
      --ctx-size 19000
      --model-draft /mnt/nvme/models/Qwen2.5-Coder-0.5B-Instruct-Q8_0.gguf
      -ngld 99
      --draft-max 16
      --draft-min 4
      --draft-p-min 0.4
      --device CUDA0
      --device-draft CUDA1
      --cache-type-k q8_0 --cache-type-v q8_0        <-- ADDED THESE FLAGS
    proxy: "http://127.0.0.1:9503"
``` 

Raw test data: 

```
# qwen-coder-32b-q4 (python, typescript, swift)
prompt eval time =      64.50 ms /     6 tokens (   10.75 ms per token,    93.02 tokens per second)
       eval time =   11264.22 ms /   900 tokens (   12.52 ms per token,    79.90 tokens per second)       
prompt eval time =      68.14 ms /     7 tokens (    9.73 ms per token,   102.73 tokens per second)
       eval time =   15766.30 ms /   859 tokens (   18.35 ms per token,    54.48 tokens per second)       
prompt eval time =      61.53 ms /     6 tokens (   10.25 ms per token,    97.51 tokens per second)
       eval time =   19349.37 ms /   903 tokens (   21.43 ms per token,    46.67 tokens per second)


# qwen-coder-32b-q4-kv (python, typescript, swift)
prompt eval time =      52.95 ms /    23 tokens (    2.30 ms per token,   434.37 tokens per second)
       eval time =   13513.06 ms /   900 tokens (   15.01 ms per token,    66.60 tokens per second)       
prompt eval time =      69.98 ms /     7 tokens (   10.00 ms per token,   100.03 tokens per second)
       eval time =   19462.99 ms /   881 tokens (   22.09 ms per token,    45.27 tokens per second)       
prompt eval time =      63.20 ms /     6 tokens (   10.53 ms per token,    94.94 tokens per second)
       eval time =   27041.86 ms /  1061 tokens (   25.49 ms per token,    39.24 tokens per second)
```

**Comment by ggerganov** [comment]
Likely the draft model should keep using F16 KV cache, otherwise it might be hitting some sub-optimal FA kernels. The change is pretty trivial in `server.cpp` - you can give it a try.

**Comment by mostlygeek** [comment]
I haven’t looked too deeply at the code before. Would you suggest that the draft model and the main model have separate settings for the kv cache type, or that the draft model always use F16?

**Comment by mostlygeek** [comment]
Thanks for the quick PR. 

Forcing the draft models cache did not make a difference. I ran my tests again with the latest commit on master and compared it to the branch: 

```
without -cache-type-[k/v], version: 4230 (0c39f44d)

prompt eval time =      52.00 ms /    23 tokens (    2.26 ms per token,   442.32 tokens per second)
       eval time =   10921.28 ms /   900 tokens (   12.13 ms per token,    82.41 tokens per second)
prompt eval time =      66.86 ms /     7 tokens (    9.55 ms per token,   104.69 tokens per second)
       eval time =   15883.79 ms /   931 tokens (   17.06 ms per token,    58.61 tokens per second)
prompt eval time =      60.51 ms /     6 tokens (   10.08 ms per token,    99.16 tokens per second)
       eval time =   21301.38 ms /  1014 tokens (   21.01 ms per token,    47.60 tokens per second)

with -cache-type-[k/v], version: 4230 (0c39f44d)

prompt eval time =      52.75 ms /    23 tokens (    2.29 ms per token,   436.03 tokens per second)
       eval time =   13575.56 ms /   900 tokens (   15.08 ms per token,    66.30 tokens per second)
prompt eval time =      68.25 ms /     7 tokens (    9.75 ms per token,   102.56 tokens per second)
       eval time =   19342.16 ms /   811 tokens (   23.85 ms per token,    41.93 tokens per second)
prompt eval time =      61.41 ms /     6 tokens (   10.24 ms per token,    97.70 tokens per second)
       eval time =   28555.61 ms /  1066 tokens (   26.79 ms per token,    37.33 tokens per second)

force KV F16 with #10586 PR, version 4225 (150d6e92):

prompt eval time =      52.89 ms /    23 tokens (    2.30 ms per token,   434.89 tokens per second)
       eval time =   13434.22 ms /   900 tokens (   14.93 ms per token,    66.99 tokens per second)
prompt eval time =      69.01 ms /     7 tokens (    9.86 ms per token,   101.44 tokens per second)
       eval time =   19370.07 ms /   895 tokens (   21.64 ms per token,    46.21 tokens per second)
prompt eval time =      63.17 ms /     6 tokens (   10.53 ms per token,    94.98 tokens per second)
       eval time =   25910.73 ms /   959 tokens (   27.02 ms per token,    37.01 tokens per second)
```

**Comment by ggerganov** [comment]
Post the results from the following commands:

```bash
llama-bench --model /mnt/nvme/models/Qwen2.5-Coder-32B-Instruct-Q4_K_M.gguf -p 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16 -n 0 -ngl 99 -fa 1

llama-bench --model /mnt/nvme/models/Qwen2.5-Coder-32B-Instruct-Q4_K_M.gguf -p 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16 -n 0 -ngl 99 -fa 1 -ctk q8_0 -ctv q8_0
```

**Comment by ggerganov** [comment]
The numbers look OK. Not sure what is causing this slow down. Does adding `"top_k": 1` to the requests help? Btw, what is the base line performance for these prompts (i.e. without draft)?

**Comment by mostlygeek** [comment]
Adding `top_k` didn't seem to make much difference

```
$ for model in "qwen-coder-32b-q4-nodraft" "qwen-coder-32b-q4" "qwen-coder-32b-q4-w-ctk"; do 
    for lang in "python" "typescript" "swift"; do
        echo "Generating Snake Game in $lang using $model"
        curl -s --url http://localhost:8080/v1/chat/completions -d "{\"messages\": [{\"role\": \"system\", \"content\": \"you only write code.\"}, {\"role\": \"user\", \"content\": \"write snake game in $lang\"}], \"temperature\": 0.1, \"top_k\": 1, \"model\":\"$model\"}" > /dev/null
        curl -s --url http://localhost:8080/logs | grep 'eval time' | tail -n 2
        echo ""
    done
done

# No Draft Model

Generating Snake Game in python using qwen-coder-32b-q4-nodraft
prompt eval time =     109.70 ms /     6 tokens (   18.28 ms per token,    54.69 tokens per second)
       eval time =   26295.40 ms /   897 tokens (   29.31 ms per token,    34.11 tokens per second)

Generating Snake Game in typescript using qwen-coder-32b-q4-nodraft
prompt eval time =      70.33 ms /     7 tokens (   10.05 ms per token,    99.53 tokens per second)
       eval time =   27249.02 ms /   927 tokens (   29.39 ms per token,    34.02 tokens per second)

Generating Snake Game in swift using qwen-coder-32b-q4-nodraft
prompt eval time =      60.11 ms /     6 tokens (   10.02 ms per token,    99.82 tokens per second)
       eval time =   28842.03 ms /   981 tokens (   29.40 ms per token,    34.01 tokens per second)

# With Draft model 

Generating Snake Game in python using qwen-coder-32b-q4
prompt eval time =      64.66 ms /    23 tokens (    2.81 ms per token,   355.69 tokens per second)
       eval time =   11050.18 ms /   900 tokens (   12.28 ms per token,    81.45 tokens per second)

Generating Snake Game in typescript using qwen-coder-32b-q4
prompt eval time =      68.51 ms /     7 tokens (    9.79 ms per token,   102.17 tokens per second)
       eval time =   15571.32 ms /   928 tokens (   16.78 ms per token,    59.60 tokens per second)

Generating Snake Game in swift using qwen-coder-32b-q4
prompt eval time =      60.99 ms /     6 tokens (   10.17 ms per token,    98.37 tokens per second)
       eval time =   22809.77 ms /   981 tokens (   23.25 ms per token,    43.01 tokens per second)

# Draft model with  --cache-type-k q8_0 --cache-type-v q8_0 

Generating Snake Game in python using qwen-coder-32b-q4-w-ctk
prompt eval time =      52.30 ms /    23 tokens (    2.27 ms per token,   439.79 tokens per second)
       eval time =   13230.13 ms /   900 tokens (   14.70 ms per token,    68.03 tokens per second)

Generating Snake Game in typescript using qwen-coder-32b-q4-w-ctk
prompt eval time =      69.92 ms /     7 tokens (    9.99 ms per token,   100.12 tokens per second)
       eval time =   19563.82 ms /   899 tokens (   21.76 ms per token,    45.95 tokens per second)

Generating Snake Game in swift using qwen-coder-32b-q4-w-ctk
prompt eval time =      60.63 ms /     6 tokens (   10.11 ms per token,    98.96 tokens per second)
       eval time =   26386.14 ms /   957 tokens (   27.57 ms per token,    36.27 tokens per second)
```

**Comment by ggerganov** [comment]
I think it should be fixed now - pull the latest https://github.com/ggerganov/llama.cpp/pull/10586 and try again.

**Comment by felladrin** [comment]
Thank you both for the investigation and resolution!

I've tested the latest changes on Qwen Coder 14B and 32B in several different queries and, in average, on a M1 Max it increased 12 t/s for the 14B model and 5 t/s for the 32B, when using  `--draft-max 16 --draft-min 4 --draft-p-min 0.4`.

**Comment by mostlygeek** [comment]
fixed in #10586.
